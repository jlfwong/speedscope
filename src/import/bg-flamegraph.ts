// https://github.com/brendangregg/FlameGraph#2-fold-stacks

import {Profile, FrameInfo, StackListProfileBuilder} from '../lib/profile'

interface BGSample {
  stack: FrameInfo[]
  duration: number
}

function parseBGFoldedStacks(contents: string): BGSample[] {
  const samples: BGSample[] = []
  contents.replace(/^(.*) (\d+)$/gm, (match: string, stack: string, n: string) => {
    samples.push({
      stack: stack.split(';').map(name => ({key: name, name: name})),
      duration: parseInt(n, 10),
    })
    return match
  })
  return samples
}

export function importFromBGFlameGraph(contents: string): Profile | null {
  const parsed = parseBGFoldedStacks(contents)
  const duration = parsed.reduce((prev: number, cur: BGSample) => prev + cur.duration, 0)
  const profile = new StackListProfileBuilder(duration)
  if (parsed.length === 0) {
    return null
  }
  for (let sample of parsed) {
    profile.appendSampleWithWeight(sample.stack, sample.duration)
  }
  return profile.build()
}
