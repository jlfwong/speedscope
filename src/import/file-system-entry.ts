// The bits of this API that we care about. This is implemented by WebKitEntry
// https://wicg.github.io/entries-api/#api-entry
export interface FileSystemDirectoryReader {
  readEntries(cb: (entries: FileSystemEntry[]) => void, error: (err: Error) => void): void
}
export interface FileSystemEntry {
  readonly isFile: boolean
  readonly isDirectory: boolean
  readonly name: string
  readonly fullPath: string
}
export interface FileSystemDirectoryEntry extends FileSystemEntry {
  createReader(): FileSystemDirectoryReader
}
export interface FileSystemFileEntry extends FileSystemEntry {
  file(cb: (file: File) => void, errCb: (err: Error) => void): void
}
