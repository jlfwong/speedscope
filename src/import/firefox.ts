import {Profile, FrameInfo, CallTreeProfileBuilder} from '../lib/profile'
import {getOrInsert} from '../lib/utils'
import {TimeFormatter} from '../lib/value-formatters'

interface Allocations {
  frames: any[]
  sites: any[]
  sizes: any[]
  timestamps: any[]
}

interface Configuration {
  allocationsMaxLogLength: number
  allocationsSampleProbability: number
  bufferSize: number
  sampleFrequency: number
  withAllocations: boolean
  withMarkers: boolean
  withMemory: boolean
  withTicks: boolean
}

interface Lib {
  arch: string
  breakpadId: string
  debugName: string
  debugPath: string
  end: any
  name: string
  offset: number
  path: string
  start: any
}

interface Meta {
  abi: string
  asyncstack: number
  debug: number
  gcpoison: number
  interval: number
  misc: string
  oscpu: string
  platform: string
  processType: number
  product: string
  shutdownTime?: any
  stackwalk: number
  startTime: number
  toolkit: string
  version: number
}

interface PausedRange {
  endTime: number
  reason: string
  startTime: number
}

type Frame = [number] | [number, number | null, number | null, number, number]

interface FrameTable {
  data: Frame[]
  /*
  schema: {
    location: 0
    implementation: 1
    optimizations: 2
    line: 3
    category: 4
  }
  */
}

interface MarkerMeta {
  category: string
  interval: string
  type: string
}
type Marker = [number, number] | [number, number, MarkerMeta]

interface Markers {
  data: Marker[]
  /*
  schema: {
    name: 0
    time: 1
    data: 2
  }
  */
}

type Sample = [number, number, number] | [number, number, number, number, number]

interface Samples {
  data: Sample[]
  /*
  schema: {
    stack: 0
    time: 1
    responsiveness: 2
    rss: 3
    uss: 4
  }
  */
}

export interface StackTable {
  data: [number | null, number][]
  /*
  schema: {
    prefix: 0
    frame: 1
  }
  */
}

export interface Thread {
  frameTable: FrameTable
  markers: Markers
  name: string
  pid: number
  processType: string
  registerTime: number
  samples: Samples
  stackTable: StackTable
  stringTable: string[]
  tid: number
  unregisterTime?: any
}

export interface FirefoxCPUProfile {
  libs: Lib[]
  meta: Meta
  pausedRanges: PausedRange[]
  processes: any[]
  threads: Thread[]
}

export interface FirefoxProfile {
  allocations: Allocations
  configuration: Configuration
  duration: number
  fileType: string
  frames: any[]
  label: string
  markers: any[]
  memory: any[]
  profile: FirefoxCPUProfile
  ticks: any[]
  version: number
}

export function importFromFirefox(firefoxProfile: FirefoxProfile): Profile {
  const cpuProfile = firefoxProfile.profile

  const thread =
    cpuProfile.threads.length === 1
      ? cpuProfile.threads[0]
      : cpuProfile.threads.filter(t => t.name === 'GeckoMain')[0]

  const frameKeyToFrameInfo = new Map<string, FrameInfo>()

  function extractStack(sample: Sample): FrameInfo[] {
    let stackFrameId: number | null = sample[0]
    const ret: number[] = []

    while (stackFrameId != null) {
      const nextStackFrame: [number | null, number] = thread.stackTable.data[stackFrameId]
      const [nextStackId, frameId] = nextStackFrame
      ret.push(frameId)
      stackFrameId = nextStackId
    }
    ret.reverse()
    return ret
      .map(f => {
        const frameData = thread.frameTable.data[f]
        const location = thread.stringTable[frameData[0]]

        const match = /(.*)\s+\((.*?):?(\d+)?\)$/.exec(location)

        if (!match) return null

        if (match[2].startsWith('resource:') || match[2] === 'self-hosted') {
          // Ignore Firefox-internals stuff
          return null
        }

        return getOrInsert(frameKeyToFrameInfo, location, () => ({
          key: location,
          name: match[1]!,
          file: match[2]!,
          line: match[3] ? parseInt(match[3]) : undefined,
        }))
      })
      .filter(f => f != null) as FrameInfo[]
  }

  const profile = new CallTreeProfileBuilder(firefoxProfile.duration)

  let prevStack: FrameInfo[] = []
  for (let sample of thread.samples.data) {
    const stack = extractStack(sample)
    const value = sample[1]

    // Find lowest common ancestor of the current stack and the previous one
    let lcaIndex = -1

    for (let i = 0; i < Math.min(stack.length, prevStack.length); i++) {
      if (prevStack[i] !== stack[i]) {
        break
      }
      lcaIndex = i
    }

    // Close frames that are no longer open
    for (let i = prevStack.length - 1; i > lcaIndex; i--) {
      profile.leaveFrame(prevStack[i], value)
    }

    for (let i = lcaIndex + 1; i < stack.length; i++) {
      profile.enterFrame(stack[i], value)
    }

    prevStack = stack
  }

  profile.setValueFormatter(new TimeFormatter('milliseconds'))
  return profile.build()
}
