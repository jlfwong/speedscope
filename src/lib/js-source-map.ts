// This file contains code to allow profiles to be remapped by JavaScript source maps.
//
// As of writing, this is using an out-of-date version of source-map, because the
// source-map library migrated to using web-assembly. This requires loading the
// web-assembly ball. The easiest way to do this is to load it from a third-party
// URL, but I want speedscope to work standalone offline. This means that the remaining
// options require some way of having a local URL that corresponds the .wasm file.
//
// Also as of writing, speedscope is bundled with Parcel v1. Trying to import
// a .wasm file in Parcel v1 tries to load the wasm module itself, which is not
// what I'm trying to do -- I want SourceMapConsumer.initialize to be the thing
// booting the WebAssembly, not Parcel itself.
//
// One way of getting around this problem is to modify the build system to
// copy the .wasm file from node_modules/source-map/lib/mappings.wasm. I could do
// this, but it's a bit of a pain.
//
// Another would be to use something like
// import("url:../node_modules/source-map/lib/mappings.wasm"), and then pass the
// resulting URL to SourceMapConsumer.initialize. This is also kind of a pain,
// because I can only do that if I upgrade to Parcel v2. Ultimately, I'd like to
// use esbuild rather than parcel at all, so for now I'm just punting on this by
// using an old-version of source-map which doesn't depend on wasm.

// This is rarely used, so let's load it async to avoid bloating the initial
// bundle.
import type {MappingItem, RawSourceMap, SourceMapConsumer} from 'source-map'
const sourceMapModule = import('source-map')

import {Frame, SymbolRemapper} from './profile'
import {findIndexBisect} from './utils'

const DEBUG = false

export async function importJavaScriptSourceMapSymbolRemapper(
  contentsString: string,
  sourceMapFileName: string,
): Promise<SymbolRemapper | null> {
  const sourceMap = await sourceMapModule

  let consumer: SourceMapConsumer | null = null
  let contents: RawSourceMap | null = null

  try {
    contents = JSON.parse(contentsString)
    consumer = new sourceMap.SourceMapConsumer(contents!)
  } catch (e) {
    return null
  }

  const mappingItems: MappingItem[] = []

  consumer.eachMapping(
    function (m: MappingItem) {
      // The sourcemap library uses 1-based line numbers, and 0-based column
      // numbers. speedscope uses 1-based line-numbers, and 1-based column
      // numbers for its in-memory representation, so we'll normalize that
      // here too.
      mappingItems.push({
        ...m,
        generatedColumn: m.generatedColumn + 1,
        originalColumn: m.originalColumn + 1,
      })
    },
    {},

    // We're going to binary search through these later, so make sure they're
    // sorted by their order in the generated file.
    sourceMap.SourceMapConsumer.GENERATED_ORDER,
  )

  const sourceMapFileNameWithoutExt = sourceMapFileName.replace(/\.[^/]*$/, '')

  return (frame: Frame) => {
    let fileMatches = false
    if (contents?.file && contents?.file === frame.file) {
      fileMatches = true
    } else if (
      ('/' + frame.file?.replace(/\.[^/]*$/, '')).endsWith('/' + sourceMapFileNameWithoutExt)
    ) {
      fileMatches = true
    }
    if (!fileMatches) {
      // The source-map doesn't apply to the file this frame is defined in.
      return null
    }

    if (frame.line == null || frame.col == null) {
      // If we don't have a line & column number for the frame, we can't
      // remap it.
      return null
    }

    // If we got here, then we hopefully have an remapping.
    //
    // Ideally, we'd look up a symbol whose generatedLine & generatedColumn
    // match what we have in our profile, but unfortunately browsers don't do
    // this.
    //
    // Browsers set the column number for a function to the index of the
    // opening paren for the argument list, rather than the beginning of the
    // index of the name.
    //
    // function alpha() { ... }
    //               ^
    //
    // const beta = function() { ... }
    //                      ^
    //
    // const gamma = () => { ... }
    //               ^
    //
    // Since we don't have the source code being profiled, we unfortunately
    // can't normalize this to set the column to the first character of the
    // actual name.
    //
    // To work around this limitation, we'll search backwards from the first
    // mapping whose generatedLine & generatedColumn are beyond the location
    // in the profile.
    let mappingIndex = findIndexBisect(mappingItems, m => {
      if (m.generatedLine > frame.line!) return true
      if (m.generatedLine < frame.line!) return false
      if (m.generatedColumn >= frame.col!) return true
      return false
    })

    if (mappingIndex === -1) {
      // There are no symbols following the given profile frame symbol, so try
      // to apply the very last mapping.
      mappingIndex = mappingItems.length - 1
    } else if (mappingIndex === 0) {
      // If the very first index in mappingItems is beyond the location in the
      // profile, it means the name we're looking for doesn't have a
      // corresponding entry in the source-map (this can happen if the
      // source-map isn't the right source-map)
      return null
    } else {
      mappingIndex--
    }

    const sourceMapItem = mappingItems[mappingIndex]
    const remappedFrameInfo: {name?: string; file?: string; line?: number; col?: number} = {}

    if (sourceMapItem.name != null) {
      remappedFrameInfo.name = sourceMapItem.name
    } else if (sourceMapItem.source != null) {
      // HACK: If the item name isn't specified, but the source is present, then
      // we're going to try to guess what the name is by using the originalLine
      // and originalColumn.

      // The second argument here is "returnNullOnMissing". Without this, it
      // throws instead of returning null.
      const content = consumer?.sourceContentFor(sourceMapItem.source, true)
      if (content) {
        const lines = content.split('\n')
        const line = lines[sourceMapItem.originalLine - 1]
        if (line) {
          // It's possible this source map entry will contain stuff other than
          // the name, so let's only consider word-ish characters that are part
          // of the prefix.
          const identifierMatch = /\w+/.exec(line.substr(sourceMapItem.originalColumn - 1))
          if (identifierMatch) {
            remappedFrameInfo.name = identifierMatch[0]
          }
        }
      }
    }

    switch (remappedFrameInfo.name) {
      case 'constructor': {
        // If the name was remapped to "constructor", then let's use the
        // original name, since "constructor" isn't very helpful.
        //
        // TODO(jlfwong): Search backwards for the class keyword and see if we
        // can guess the right name.
        remappedFrameInfo.name = frame.name + ' constructor'
        break
      }

      case 'function': {
        // If the name is just "function", it probably means we either messed up
        // the remapping, or that we matched an anonymous function. In either
        // case, this isn't helpful, so put this back.
        remappedFrameInfo.name = frame.name
        break
      }

      case 'const':
      case 'export': {
        // If we got this, we probably just did a bad job leveraging the hack
        // looking through the source code. Let's fall-back to whatever the
        // original name was.
        remappedFrameInfo.name = frame.name
        break
      }
    }

    if (remappedFrameInfo.name && frame.name.includes(remappedFrameInfo.name)) {
      // If the remapped name is a substring of the original name, the original
      // name probably contains more useful information. In that case, just use
      // the original name instead.
      //
      // This can happen, for example, when remapping method names. If a
      // call stack says the symbol name is "n.zap" and we remapped it to a
      // function just called "zap", we might as well use the original name
      // instead.
      remappedFrameInfo.name = frame.name
    }

    if (sourceMapItem.source != null) {
      remappedFrameInfo.file = sourceMapItem.source
      remappedFrameInfo.line = sourceMapItem.originalLine
      remappedFrameInfo.col = sourceMapItem.originalColumn
    }

    if (DEBUG) {
      console.groupCollapsed(`Remapping "${frame.name}" -> "${remappedFrameInfo.name}"`)
      console.log('before', {...frame})
      console.log('item @ index', sourceMapItem)
      console.log('item @ index + 1', mappingItems[mappingIndex + 1])
      console.log('after', remappedFrameInfo)
      console.groupEnd()
    }

    return remappedFrameInfo
  }
}
