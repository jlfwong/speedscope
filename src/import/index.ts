import {Profile, ProfileGroup} from '../lib/profile'
import {FileSystemDirectoryEntry} from './file-system-entry'

import {importFromChromeCPUProfile, importFromChromeTimeline} from './chrome'
import {importFromStackprof} from './stackprof'
import {importFromInstrumentsDeepCopy, importFromInstrumentsTrace} from './instruments'
import {importFromBGFlameGraph} from './bg-flamegraph'
import {importFromFirefox} from './firefox'
import {importSpeedscopeProfiles} from '../lib/file-format'
import {importFromV8ProfLog} from './v8proflog'
import {importFromLinuxPerf} from './linux-tools-perf'

export async function importProfileGroup(
  fileName: string,
  contents: string,
): Promise<ProfileGroup | null> {
  const profileGroup = await _importProfileGroup(fileName, contents)
  if (profileGroup) {
    if (!profileGroup.name) {
      profileGroup.name = fileName
    }
    for (let profile of profileGroup.profiles) {
      if (profile && !profile.getName()) {
        profile.setName(fileName)
      }
    }
    return profileGroup
  }
  return null
}

function toGroup(profile: Profile | null): ProfileGroup | null {
  if (!profile) return null
  return {name: profile.getName(), indexToView: 0, profiles: [profile]}
}

async function _importProfileGroup(
  fileName: string,
  contents: string,
): Promise<ProfileGroup | null> {
  // First pass: Check known file format names to infer the file type
  if (fileName.endsWith('.speedscope.json')) {
    console.log('Importing as speedscope json file')
    return importSpeedscopeProfiles(JSON.parse(contents))
  } else if (fileName.endsWith('.cpuprofile')) {
    console.log('Importing as Chrome CPU Profile')
    return toGroup(importFromChromeCPUProfile(JSON.parse(contents)))
  } else if (fileName.endsWith('.chrome.json') || /Profile-\d{8}T\d{6}/.exec(fileName)) {
    console.log('Importing as Chrome Timeline')
    return toGroup(importFromChromeTimeline(JSON.parse(contents)))
  } else if (fileName.endsWith('.stackprof.json')) {
    console.log('Importing as stackprof profile')
    return toGroup(importFromStackprof(JSON.parse(contents)))
  } else if (fileName.endsWith('.instruments.txt')) {
    console.log('Importing as Instruments.app deep copy')
    return toGroup(importFromInstrumentsDeepCopy(contents))
  } else if (fileName.endsWith('.linux-perf.txt')) {
    console.log('Importing as output of linux perf script')
    return importFromLinuxPerf(contents)
  } else if (fileName.endsWith('.collapsedstack.txt')) {
    console.log('Importing as collapsed stack format')
    return toGroup(importFromBGFlameGraph(contents))
  } else if (fileName.endsWith('.v8log.json')) {
    console.log('Importing as --prof-process v8 log')
    return toGroup(importFromV8ProfLog(JSON.parse(contents)))
  }

  // Second pass: Try to guess what file format it is based on structure
  let parsed: any
  try {
    parsed = JSON.parse(contents)
  } catch (e) {}
  if (parsed) {
    if (parsed['$schema'] === 'https://www.speedscope.app/file-format-schema.json') {
      console.log('Importing as speedscope json file')
      return importSpeedscopeProfiles(JSON.parse(contents))
    } else if (parsed['systemHost'] && parsed['systemHost']['name'] == 'Firefox') {
      console.log('Importing as Firefox profile')
      return toGroup(importFromFirefox(parsed))
    } else if (Array.isArray(parsed) && parsed[parsed.length - 1].name === 'CpuProfile') {
      console.log('Importing as Chrome CPU Profile')
      return toGroup(importFromChromeTimeline(parsed))
    } else if ('nodes' in parsed && 'samples' in parsed && 'timeDeltas' in parsed) {
      console.log('Importing as Chrome Timeline')
      return toGroup(importFromChromeCPUProfile(parsed))
    } else if ('mode' in parsed && 'frames' in parsed) {
      console.log('Importing as stackprof profile')
      return toGroup(importFromStackprof(parsed))
    } else if ('code' in parsed && 'functions' in parsed && 'ticks' in parsed) {
      console.log('Importing as --prof-process v8 log')
      return toGroup(importFromV8ProfLog(parsed))
    }
  } else {
    // Format is not JSON

    // If the first line contains "Symbol Name", preceded by a tab, it's probably
    // a deep copy from OS X Instruments.app
    if (/^[\w \t\(\)]*\tSymbol Name/.exec(contents)) {
      console.log('Importing as Instruments.app deep copy')
      return toGroup(importFromInstrumentsDeepCopy(contents))
    }

    // If every line ends with a space followed by a number, it's probably
    // the collapsed stack format.
    const lineCount = contents.split(/\n/).length
    if (lineCount >= 1 && lineCount === contents.split(/ \d+\n/).length) {
      console.log('Importing as collapsed stack format')
      return toGroup(importFromBGFlameGraph(contents))
    }

    const fromLinuxPerf = importFromLinuxPerf(contents)
    if (fromLinuxPerf) {
      console.log('Importing from linux perf script output')
      return fromLinuxPerf
    }
  }

  // Unrecognized format
  return null
}

export async function importFromFileSystemDirectoryEntry(entry: FileSystemDirectoryEntry) {
  return importFromInstrumentsTrace(entry)
}
