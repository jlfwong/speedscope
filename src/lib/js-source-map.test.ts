import * as fs from 'fs'
import * as path from 'path'
import {importProfilesFromArrayBuffer} from '../import'
import {importJavaScriptSourceMapSymbolRemapper} from './js-source-map'
import {Frame} from './profile'

async function checkSourceMapApplication(pathToProfile: string, pathToSourceMap: string) {
  const buffer = fs.readFileSync(pathToProfile)
  const arrayBuffer = buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength)
  const profileGroup = await importProfilesFromArrayBuffer(
    path.basename(pathToProfile),
    arrayBuffer,
  )

  if (!profileGroup) {
    fail('Failed to extract profile')
    return
  }

  const sourceMapFileName = path.basename(pathToSourceMap)
  const remapper = await importJavaScriptSourceMapSymbolRemapper(
    fs.readFileSync(pathToSourceMap, 'utf-8'),
    sourceMapFileName,
  )

  if (!remapper) {
    fail('Failed to extract sourcemap')
    return
  }

  const key: (f: {name?: string; file?: string; line?: number; col?: number}) => string = f => {
    return `${f.name} @ ${f.file}:${f.line}:${f.col}`
  }

  const frames: Frame[] = []

  profileGroup.profiles[profileGroup.indexToView].forEachFrame(f => {
    frames.push(f)
  })

  frames.sort((a, b) => (a.key < b.key ? -1 : 1))

  const remappedFrames: string[] = []
  frames.forEach(f => {
    const remapped = remapper(f)
    if (!remapped) return
    remappedFrames.push(`(${key({...f, ...remapped})}) <- (${key(f)})`)
  })

  expect(remappedFrames).toMatchSnapshot()
}

test('source-map remapping of chrome-85-webpack', async () => {
  await checkSourceMapApplication(
    './sample/profiles/source-maps/chrome-85-webpack.json',
    './sample/profiles/source-maps/webpack/typescript-source-map-test.js.map',
  )
})

test('source-map remapping of firefox-79-webpack', async () => {
  await checkSourceMapApplication(
    './sample/profiles/source-maps/firefox-79-webpack.json',
    './sample/profiles/source-maps/webpack/typescript-source-map-test.js.map',
  )
})

test('source-map remapping of safari-13-webpack', async () => {
  await checkSourceMapApplication(
    './sample/profiles/source-maps/safari-13-webpack.json',
    './sample/profiles/source-maps/webpack/typescript-source-map-test.js.map',
  )
})

test('source-map remapping of chrome-85-esbuild', async () => {
  await checkSourceMapApplication(
    './sample/profiles/source-maps/chrome-85-esbuild.json',
    './sample/profiles/source-maps/esbuild/typescript-source-map-test.js.map',
  )
})

test('source-map remapping of chrome-85-parcel', async () => {
  await checkSourceMapApplication(
    './sample/profiles/source-maps/chrome-85-parcel.json',
    './sample/profiles/source-maps/parcel/typescript-source-map-test.js.map',
  )
})
