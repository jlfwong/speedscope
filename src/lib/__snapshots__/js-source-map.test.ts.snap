// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source-map remapping of chrome-85-esbuild 1`] = `
Array [
  "((anonymous typescript-source-map-test.js:1) @ alpha.ts:5:5) <- ((anonymous typescript-source-map-test.js:1) @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/esbuild/typescript-source-map-test.js:1:170)",
  "(e @ beta.ts:3:8) <- (e @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/esbuild/typescript-source-map-test.js:1:75)",
  "(get floop @ kludge.ts:13:7) <- (get floop @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/esbuild/typescript-source-map-test.js:1:276)",
  "(l constructor @ kludge.ts:4:3) <- (l @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/esbuild/typescript-source-map-test.js:1:226)",
  "(m @ delta.ts:3:22) <- (m @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/esbuild/typescript-source-map-test.js:1:119)",
  "(r @ alpha.ts:4:8) <- (r @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/esbuild/typescript-source-map-test.js:1:158)",
  "(gamma @ gamma.ts:1:14) <- (t @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/esbuild/typescript-source-map-test.js:1:15)",
  "(zap @ kludge.ts:9:3) <- (zap @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/esbuild/typescript-source-map-test.js:1:260)",
]
`;

exports[`source-map remapping of chrome-85-parcel 1`] = `
Array [
  "((anonymous typescript-source-map-test.js:8) @ ../alpha.ts:5:5) <- ((anonymous typescript-source-map-test.js:8) @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:8:152)",
  "(e @ ../gamma.ts:1:22) <- (e @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:2:104)",
  "(get @ ../kludge.ts:13:3) <- (get @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:10:284)",
  "(o.zap @ ../kludge.ts:9:3) <- (o.zap @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:10:211)",
  "(o constructor @ ../kludge.ts:4:3) <- (o @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:10:139)",
  "(parcelRequire.CIJJ../alpha @ ../alpha.ts:5:5) <- (parcelRequire.CIJJ../alpha @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:9:55)",
  "(parcelRequire.EJAe../gamma @ ../gamma.ts:1:8) <- (parcelRequire.EJAe../gamma @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:3:23)",
  "(parcelRequire.NDR3../beta @ ../delta.ts:3:8) <- (parcelRequire.NDR3../beta @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:7:39)",
  "(parcelRequire.xEzo../kludge @ ../kludge.ts:3:1) <- (parcelRequire.xEzo../kludge @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:11:39)",
  "(beta @ ../beta.ts:3:17) <- (r @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:4:119)",
  "(r @ ../delta.ts:3:22) <- (r @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:6:125)",
  "(alpha @ ../alpha.ts:4:17) <- (t @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/parcel/typescript-source-map-test.js:8:140)",
]
`;

exports[`source-map remapping of chrome-85-webpack 1`] = `
Array [
  "((anonymous typescript-source-map-test.js:1) @ webpack://speedscope-sourcemap-test-project/alpha.ts:5:4) <- ((anonymous typescript-source-map-test.js:1) @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:181)",
  "(get @ webpack://speedscope-sourcemap-test-project/kludge.ts:13:3) <- (get @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:368)",
  "(n.zap @ webpack://speedscope-sourcemap-test-project/kludge.ts:9:3) <- (n.zap @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:305)",
  "(n constructor @ webpack://speedscope-sourcemap-test-project/kludge.ts:4:3) <- (n @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:243)",
  "(n @ webpack://speedscope-sourcemap-test-project/gamma.ts:1:22) <- (n @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:34)",
  "(beta @ webpack://speedscope-sourcemap-test-project/beta.ts:3:17) <- (o @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:88)",
  "(r @ webpack://speedscope-sourcemap-test-project/delta.ts:3:22) <- (r @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:130)",
  "(alpha @ webpack://speedscope-sourcemap-test-project/alpha.ts:4:17) <- (t @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:169)",
]
`;

exports[`source-map remapping of firefox-79-webpack 1`] = `
Array [
  "(get @ webpack://speedscope-sourcemap-test-project/kludge.ts:13:3) <- (get @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:368)",
  "(n constructor @ webpack://speedscope-sourcemap-test-project/kludge.ts:4:3) <- (n @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:243)",
  "(n @ webpack://speedscope-sourcemap-test-project/gamma.ts:1:22) <- (n @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:34)",
  "(n.prototype.zap @ webpack://speedscope-sourcemap-test-project/kludge.ts:9:3) <- (n.prototype.zap @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:305)",
  "(beta @ webpack://speedscope-sourcemap-test-project/beta.ts:3:17) <- (o @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:88)",
  "(r @ webpack://speedscope-sourcemap-test-project/delta.ts:3:22) <- (r @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:130)",
  "(alpha @ webpack://speedscope-sourcemap-test-project/alpha.ts:4:17) <- (t @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:169)",
  "(t/< @ webpack://speedscope-sourcemap-test-project/alpha.ts:5:4) <- (t/< @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:181)",
]
`;

exports[`source-map remapping of safari-13-webpack 1`] = `
Array [
  "((anonymous typescript-source-map-test.js:1) @ webpack://speedscope-sourcemap-test-project/alpha.ts:5:4) <- ((anonymous typescript-source-map-test.js:1) @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:181)",
  "(zap @ webpack://speedscope-sourcemap-test-project/kludge.ts:9:3) <- ((anonymous typescript-source-map-test.js:1) @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:305)",
  "(get @ webpack://speedscope-sourcemap-test-project/kludge.ts:13:3) <- (get @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:368)",
  "(n constructor @ webpack://speedscope-sourcemap-test-project/kludge.ts:4:3) <- (n @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:243)",
  "(n @ webpack://speedscope-sourcemap-test-project/gamma.ts:1:22) <- (n @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:34)",
  "(beta @ webpack://speedscope-sourcemap-test-project/beta.ts:3:17) <- (o @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:88)",
  "(r @ webpack://speedscope-sourcemap-test-project/delta.ts:3:22) <- (r @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:130)",
  "(alpha @ webpack://speedscope-sourcemap-test-project/alpha.ts:4:17) <- (t @ file:///Users/jlfwong/code/speedscope/sample/programs/javascript/sourcemaps/dist/webpack/typescript-source-map-test.js:1:169)",
]
`;
